kubectl-plugin/pkg/cmd/submit/submit.go

```diff
@@ -0,0 +1,537 @@
+		}
+
+		submissionMode, ok := options.RayJob.Object["spec"].(map[string]interface{})["submissionMode"]
+		if !ok {
+			return fmt.Errorf("RayJob does not have `submissionMode` field set")
+		}
+		if submissionMode != nil {
+			// Currently using string since latest release does not have `rayv1api.UserMode` yet
+			if submissionMode != "UserMode" {
+				return fmt.Errorf("Submission mode of the Ray Job is not supported")
+			}
+		} else {
+			return fmt.Errorf("Submission mode must be set to rayv1api.UserMode or `UserMode`")
+		}
+	}
+
+	return nil
+}
+
+func (options *SubmitJobOptions) Run(ctx context.Context, factory cmdutil.Factory) error {
+	k8sClients, err := client.NewClient(factory)
+	if err != nil {
+		return fmt.Errorf("failed to initialize clientset: %w", err)
+	}
+
+	if options.RayJob != nil {
+		// createdRayJob, err = k8sClients.CreateRayCustomResource(ctx, util.RayJob, options.configFlags.Namespace, unstructuredRayjob)
+		options.RayJob, err = k8sClients.DynamicClient().Resource(util.RayJobGVR).Namespace(*options.configFlags.Namespace).Create(ctx, options.RayJob, v1.CreateOptions{})
+		if err != nil {
+			return fmt.Errorf("Error when creating RayJob CR: %w", err)
+		}
+		fmt.Printf("Submitted RayJob %s.\n", options.RayJob.GetName())
+	}
+
+	if options.RayJob != nil && len(options.RayJob.GetName()) > 0 {
+		// Add timeout?
+		for options.RayJob.Object["status"] == nil {
+			options.RayJob, err = k8sClients.DynamicClient().Resource(util.RayJobGVR).Namespace(*options.configFlags.Namespace).Get(ctx, options.RayJob.GetName(), v1.GetOptions{})
+			if err != nil {
+				return fmt.Errorf("Failed to get Ray Job status")
+			}
+		}
+		clusterName, ok := options.RayJob.Object["status"].(map[string]interface{})["rayClusterName"].(string)
+		if !ok {
+			return fmt.Errorf("Unable to find ray cluster status")
+		}
+		if len(clusterName) == 0 {
+			return fmt.Errorf("No cluster name available even after status of Ray Job is set")
+		}
+		options.cluster = clusterName
+	} else {
+		return fmt.Errorf("Unknown cluster and did not provide Ray Job. One of the fields must be set")
+	}
+
+	// Wait til the cluster is ready
+	var ready bool
+	maxSeconds := 120.0
+	startTime := time.Now()
+	currTime := time.Now()
+	fmt.Printf("Waiting for Cluster Ready\n")
+	for !ready && currTime.Sub(startTime).Seconds() <= maxSeconds {
+		time.Sleep(2 * time.Second)
+		fmt.Printf("Checking Cluster Status for cluster %s...\n", options.cluster)
+		currCluster, err := k8sClients.DynamicClient().Resource(util.RayClusterGVR).Namespace(*options.configFlags.Namespace).Get(ctx, options.cluster, v1.GetOptions{})
+		if err != nil {
+			return fmt.Errorf("Failed to get cluster information with error: %w", err)
+		}
+		ready, err = extractClusterStatus(currCluster)
+		if err != nil {
+			return fmt.Errorf("Failed to get cluster status: %w", err)
+		}
+		currTime = time.Now()
+	}
+
+	if !ready {
+		if options.RayJob != nil {
```

Comment:
Is it valid to not specify a RayJob? If not, I suggest checking for `options.RayJob != nil` once and returning an error early and removing all the `options.RayJob != nil` checks everywhere else

Reply:
I think it should be valid to not specify a ray job. I'm thinking of the case where user already have a raycluster ready and just want to submit the job.

Reply:
I'm leaning towards making this specific to RayJob. For the first case you mentioned for submitting a job to a RayCluster, that's not any different from just using `ray job submit`. 

What do you think @MortalHappiness @kevin85421

Reply:
I also think we should only support submit to RayJob for this command, because if users already have a RayCluster, they can use `ray session` followed by `ray job submit` CLI directly.

Reply:
I will go with making this specific to RayJob. I agree that it shouldn't be used as another way to submit a ray job/a way to pass command to ray.